src/main/scala/lms/core/stub.scala

```diff
@@ -132,6 +132,61 @@ trait Base extends EmbeddedControls with OverloadHack with lms.util.ClosureCompa
 
   implicit def unit[T:Manifest](x: T): Rep[T] = Wrap[T](Backend.Const(x))
 
+
+  class TOP(val x: Backend.Exp) {
+    def withSource(pos: SourceContext) = { Adapter.sourceMap(x) = pos; this }
```

Comment:
There is some redundancy of `withSource`, `withType`, and `with_`, but none of them override the other. it would be good to abstract the code a little bit using self type or traditional OO technique (such as F-bounded polymorphism) to express the invariant, like `withSource` should return the same type of this class, then it seems we can get rid of some of the castings.

Reply:
maybe have a chat about this? @Kraks